File: clients/python/astra_sdk/runner.py
"""LocalRunner — serve a Astra-deployed artifact on your own hardware.

    from astra_sdk import LocalRunner

    runner = LocalRunner.from_deployment(
        base_url="https://app.example.com",
        deployment_id="dep_ab12cd34ef",
        api_key="astra_sk_live_…",
    )
    out = runner.run({"input": my_numpy_array})    # local onnxruntime inference
    runner.close()

The artifact is pulled once via the API-key-authed
GET /api/v1/artifacts/{deployment_id} and cached on disk keyed by its sha256,
so restarts don't re-download. Every run() is measured (pre/infer/post) and
shipped to the Astra dashboard by the background TelemetryReporter, together
with periodic system snapshots and windowed input/output stats — the same
closed loop hosted serving gets, but on your hardware.

Requires the [serve] extra:  pip install 'astra-ai-sdk[serve]'
"""

from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any

from ._http import HttpSession, resolve_base_url
from .telemetry import TelemetryReporter

_DEFAULT_CACHE = "~/.cache/astra"


class RunnerError(Exception):
    pass


def _require_serve_extra():
    try:
        import numpy
        import onnxruntime
    except ImportError as exc:  # pragma: no cover - exercised only without extra
        raise RunnerError(
            "Local serving needs onnxruntime + numpy — install the extra: "
            "pip install 'astra-ai-sdk[serve]'"
        ) from exc
    return numpy, onnxruntime


class LocalRunner:
    """Local ONNX serving with built-in telemetry."""

    def __init__(
        self,
        model_path: str,
        *,
        reporter: TelemetryReporter | None = None,
        providers: list[str] | None = None,
    ) -> None:
        np, ort = _require_serve_extra()
        self._np = np
        self._reporter = reporter
        self._session = ort.InferenceSession(
            model_path,
            providers=providers or ort.get_available_providers(),
        )
        self._input_meta = self._session.get_inputs()
        self._output_names = [o.name for o in self._session.get_outputs()]
        self.model_path = model_path

    @property
    def active_provider(self) -> str:
        """The execution provider ORT actually bound (first in get_providers)."""
        try:
            providers = self._session.get_providers()
            return providers[0] if providers else ""
        except Exception:  # noqa: BLE001
            return ""

    # ── construction ─────────────────────────────────────────────────────────

    @classmethod
    def from_deployment(
        cls,
        deployment_id: str,
        api_key: str,
        *,
        base_url: str | None = None,
        cache_dir: str = _DEFAULT_CACHE,
        report_telemetry: bool = True,
        providers: list[str] | None = None,
        timeout: float = 60.0,
    ) -> "LocalRunner":
        """Pull (or reuse) the deployed artifact and build a runner for it.

        base_url defaults to the hosted Astra origin (override with the
        ASTRA_BASE_URL env var or the base_url keyword)."""
        _require_serve_extra()
        base = resolve_base_url(base_url)
        path = pull_artifact(
            deployment_id, api_key, base_url=base,
            cache_dir=cache_dir, timeout=timeout,
        )
        from . import __version__

        # Build the session first so the reporter can record the provider ORT
        # actually selected — not just the first one available on the box.
        runner = cls(str(path), providers=providers)
        runner._reporter = TelemetryReporter(
            base, deployment_id, api_key,
            sdk_version=__version__, enabled=report_telemetry,
            active_provider=runner.active_provider,
        )
        return runner

    @classmethod
    def from_file(
        cls,
        model_path: str,
        *,
        providers: list[str] | None = None,
        report_telemetry: bool = False,
        deployment_id: str | None = None,
        api_key: str | None = None,
        base_url: str | None = None,
    ) -> "LocalRunner":
        """Serve a model file you ALREADY have on disk — e.g. the one from the
        SDK Hub "Download Artifact" button. No deployment / network needed.
        Telemetry is OFF unless you also pass deployment_id + api_key (then local
        runs still report to that deployment)."""
        _require_serve_extra()
        runner = cls(model_path, providers=providers)
        want = report_telemetry or bool(deployment_id and api_key)
        if want and deployment_id and api_key:
            from . import __version__

            runner._reporter = TelemetryReporter(
                resolve_base_url(base_url), deployment_id, api_key,
                sdk_version=__version__, enabled=True,
                active_provider=runner.active_provider,
            )
        return runner

    # ── serving ──────────────────────────────────────────────────────────────

    def run(
        self,
        inputs: dict[str, Any] | None = None,
        *,
        region: str = "local",
    ) -> dict[str, Any]:
        """One inference. `inputs` maps input name → ndarray/nested list; pass
        None to synthesize a random valid probe (smoke tests / benchmarks)."""
        np = self._np
        t0 = time.perf_counter()
        error_code: str | None = None
        feeds: dict[str, Any] = {}
        try:
            feeds = self._prepare(inputs)
            pre_ms = (time.perf_counter() - t0) * 1000

            t1 = time.perf_counter()
            raw = self._session.run(self._output_names, feeds)
            infer_ms = (time.perf_counter() - t1) * 1000

            t2 = time.perf_counter()
            outputs = [
                {"name": n, "shape": list(np.asarray(o).shape)}
                for n, o in zip(self._output_names, raw)
            ]
            first = raw[0] if raw else None
            if self._reporter is not None:
                self._reporter.observe(feeds, first)
            post_ms = (time.perf_counter() - t2) * 1000
        except Exception as exc:
            error_code = "bad_input" if isinstance(exc, (ValueError, KeyError)) else "inference_error"
            if self._reporter is not None:
                self._reporter.record_event(
                    latency_ms=0.0, success=False, error_code=error_code,
                    batch_size=self._batch_of(feeds), region=region,
                    input_sig=self._signature(feeds),
                )
            raise

        if self._reporter is not None:
            self._reporter.record_event(
                latency_ms=infer_ms, pre_ms=pre_ms, post_ms=post_ms,
                success=True, batch_size=self._batch_of(feeds), region=region,
                input_sig=self._signature(feeds),
            )
        return {
            "latencyMs": round(infer_ms, 3),
            "preMs": round(pre_ms, 3),
            "postMs": round(post_ms, 3),
            "outputs": outputs,
            "raw": raw,
        }

    def _prepare(self, inputs: dict[str, Any] | None) -> dict[str, Any]:
        np = self._np
        feeds: dict[str, Any] = {}
        if inputs is None:
            for meta in self._input_meta:
                shape = [d if isinstance(d, int) and d > 0 else 1 for d in meta.shape]
                feeds[meta.name] = np.random.default_rng().standard_normal(
                    shape).astype(np.float32)
            return feeds
        for meta in self._input_meta:
            if meta.name not in inputs:
                raise ValueError(f"missing input '{meta.name}'")
            arr = np.asarray(inputs[meta.name])
            if arr.dtype.kind == "f":
                arr = arr.astype(np.float32)
            feeds[meta.name] = arr
        return feeds

    def _batch_of(self, feeds: dict[str, Any]) -> int:
        for arr in feeds.values():
            shape = getattr(arr, "shape", None)
            if shape:
                return int(shape[0])
        return 1

    def _signature(self, feeds: dict[str, Any]) -> str | None:
        parts = []
        for name, arr in feeds.items():
            shape = getattr(arr, "shape", ())
            dtype = getattr(arr, "dtype", "")
            parts.append(f"{name}:{'x'.join(str(d) for d in shape)}:{dtype}")
        return ";".join(parts) or None

    # ── lifecycle ────────────────────────────────────────────────────────────

    def close(self) -> None:
        if self._reporter is not None:
            self._reporter.close()

    def __enter__(self) -> "LocalRunner":
        return self

    def __exit__(self, *_exc: object) -> None:
        self.close()


def pull_artifact(
    deployment_id: str,
    api_key: str,
    *,
    base_url: str | None = None,
    cache_dir: str = _DEFAULT_CACHE,
    timeout: float = 60.0,
) -> Path:
    """Download the deployed artifact (sha256-cached under cache_dir).

    base_url defaults to the hosted Astra origin (override with ASTRA_BASE_URL)."""
    http = HttpSession(resolve_base_url(base_url), api_key, timeout=timeout)
    try:
        info = http.request(
            "GET", f"/api/v1/artifacts/{deployment_id}/info").json()
        sha = info["sha256"]
        root = Path(cache_dir).expanduser() / deployment_id
        root.mkdir(parents=True, exist_ok=True)
        target = root / f"{sha}{Path(info['fileName']).suffix or '.onnx'}"
        if target.exists() and target.stat().st_size == info["sizeBytes"]:
            return target
        resp = http.request("GET", f"/api/v1/artifacts/{deployment_id}")
        tmp = target.with_suffix(".part")
        tmp.write_bytes(resp.content)
        tmp.replace(target)
        (root / "meta.json").write_text(json.dumps(info, indent=2))
        return target
    finally:
        http.close()

File: clients/node/src/runner.ts
// LocalRunner — serve a Astra-deployed artifact on your own hardware.
// Mirrors clients/python/astra_sdk/runner.py.
//
//     import { LocalRunner } from "astra-ai-sdk";
//
//     const runner = await LocalRunner.fromDeployment({
//       baseUrl: "https://app.example.com",
//       deploymentId: "dep_ab12cd34ef",
//       apiKey: "astra_sk_live_…",
//     });
//     const out = await runner.run({ input: { data: myFloats, dims: [1, 3, 224, 224] } });
//     await runner.close();
//
// The artifact is pulled once via the API-key-authed
// GET /api/v1/artifacts/{deployment_id} and cached on disk keyed by its sha256,
// so restarts don't re-download. Every run() is measured (pre/infer/post) and
// shipped to the Astra dashboard by the background TelemetryReporter.
//
// Requires onnxruntime-node:  npm i onnxruntime-node

import { existsSync, statSync } from "node:fs";
import { mkdir, rename, writeFile } from "node:fs/promises";
import { createRequire } from "node:module";
import { homedir } from "node:os";
import { extname, join } from "node:path";

import { HttpSession, resolveBaseUrl } from "./http.js";
import { TelemetryReporter } from "./telemetry.js";
import {
  batchOf,
  buildTensor,
  type InputMeta,
  randomProbe,
  signatureOf,
} from "./tensor.js";
import { VERSION } from "./version.js";
import type {
  OrtInferenceSession,
  OrtModule,
  OrtTensor,
  OrtValueMetadata,
} from "./ort.js";
import type { RunInput, RunOutput } from "./types.js";
import type { Sampleable } from "./stats.js";

const DEFAULT_CACHE = "~/.cache/astra";

export class RunnerError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RunnerError";
  }
}

/** Thrown for caller mistakes (missing input) so run() classifies them as
 *  "bad_input" — matching the Python SDK's ValueError/KeyError handling. */
class BadInputError extends Error {}

let cachedOrt: OrtModule | null = null;

/** Lazily load onnxruntime-node; throw an actionable RunnerError if absent. */
export async function requireServeExtra(): Promise<OrtModule> {
  if (cachedOrt) return cachedOrt;
  try {
    // Non-literal specifier: keeps TypeScript from trying to resolve the
    // optional native module at build time (it's loaded only at runtime).
    const spec = "onnxruntime-node";
    const mod = (await import(spec)) as { default?: OrtModule } & OrtModule;
    cachedOrt = (mod.default ?? mod) as OrtModule;
    return cachedOrt;
  } catch {
    throw new RunnerError(
      "Local serving needs onnxruntime-node — install it: npm i onnxruntime-node",
    );
  }
}

function ortVersion(): string {
  try {
    const require = createRequire(import.meta.url);
    const pkg = require("onnxruntime-node/package.json") as { version?: string };
    return pkg.version ?? "";
  } catch {
    return "";
  }
}

function expandHome(dir: string): string {
  if (dir === "~") return homedir();
  if (dir.startsWith("~/") || dir.startsWith("~\\")) {
    return join(homedir(), dir.slice(2));
  }
  return dir;
}

export interface PullArtifactOptions {
  deploymentId: string;
  apiKey: string;
  /** Optional — defaults to the hosted Astra origin (or ASTRA_BASE_URL). */
  baseUrl?: string;
  cacheDir?: string;
  timeout?: number;
}

/** Download the deployed artifact (sha256-cached under cacheDir). */
export async function pullArtifact(opts: PullArtifactOptions): Promise<string> {
  const { deploymentId, apiKey } = opts;
  const cacheDir = opts.cacheDir ?? DEFAULT_CACHE;
  const http = new HttpSession(resolveBaseUrl(opts.baseUrl), apiKey, {
    timeout: opts.timeout ?? 60,
  });
  try {
    const infoResp = await http.request("GET", `/api/v1/artifacts/${deploymentId}/info`);
    const info = (await infoResp.json()) as {
      fileName: string;
      sizeBytes: number;
      sha256: string;
    };
    const root = join(expandHome(cacheDir), deploymentId);
    await mkdir(root, { recursive: true });
    const suffix = extname(info.fileName) || ".onnx";
    const target = join(root, `${info.sha256}${suffix}`);
    if (existsSync(target) && statSync(target).size === info.sizeBytes) {
      return target;
    }
    const resp = await http.request("GET", `/api/v1/artifacts/${deploymentId}`);
    const buf = Buffer.from(await resp.arrayBuffer());
    const tmp = `${target}.part`;
    await writeFile(tmp, buf);
    await rename(tmp, target);
    await writeFile(join(root, "meta.json"), JSON.stringify(info, null, 2));
    return target;
  } finally {
    http.close();
  }
}

export interface FromDeploymentOptions {
  deploymentId: string;
  apiKey: string;
  /** Optional — defaults to the hosted Astra origin (or ASTRA_BASE_URL). */
  baseUrl?: string;
  cacheDir?: string;
  reportTelemetry?: boolean;
  providers?: string[];
  timeout?: number;
}

/** Local ONNX serving with built-in telemetry. */
export class LocalRunner {
  private reporter: TelemetryReporter | null = null;
  private readonly metas: InputMeta[];
  private readonly activeProviderName: string;

  private constructor(
    private readonly ort: OrtModule,
    private readonly session: OrtInferenceSession,
    activeProvider: string,
  ) {
    this.activeProviderName = activeProvider;
    this.metas = this.readInputMetas();
  }

  static async fromDeployment(opts: FromDeploymentOptions): Promise<LocalRunner> {
    const ort = await requireServeExtra();
    const base = resolveBaseUrl(opts.baseUrl);
    const path = await pullArtifact({
      baseUrl: base,
      deploymentId: opts.deploymentId,
      apiKey: opts.apiKey,
      cacheDir: opts.cacheDir,
      timeout: opts.timeout ?? 60,
    });
    const session = await ort.InferenceSession.create(
      path,
      opts.providers ? { executionProviders: opts.providers } : undefined,
    );
    // onnxruntime-node does not reliably expose the EP it actually bound, so we
    // record the requested provider (or CPU) rather than fabricate one.
    const active = opts.providers?.[0] ?? "CPUExecutionProvider";
    const runner = new LocalRunner(ort, session, active);
    runner.reporter = new TelemetryReporter(base, opts.deploymentId, opts.apiKey, {
      sdkVersion: VERSION,
      enabled: opts.reportTelemetry ?? true,
      activeProvider: active,
      ortVersion: ortVersion(),
      availableProviders: opts.providers ?? [active],
    });
    return runner;
  }

  /**
   * Serve a model file you ALREADY have on disk — e.g. the one from the SDK Hub
   * "Download Artifact" button, or an artifact committed next to your code. No
   * deployment / network needed. Telemetry is OFF unless you also pass a
   * deployment id + API key (then local runs still report to that deployment).
   */
  static async fromFile(
    modelPath: string,
    opts: {
      providers?: string[];
      deploymentId?: string;
      apiKey?: string;
      baseUrl?: string;
      reportTelemetry?: boolean;
    } = {},
  ): Promise<LocalRunner> {
    const ort = await requireServeExtra();
    const session = await ort.InferenceSession.create(
      modelPath,
      opts.providers ? { executionProviders: opts.providers } : undefined,
    );
    const active = opts.providers?.[0] ?? "CPUExecutionProvider";
    const runner = new LocalRunner(ort, session, active);
    // Report telemetry only when a deployment is supplied — there's nowhere to
    // send it otherwise. Defaults on in that case, off for a bare file.
    const wantTelemetry =
      opts.reportTelemetry ?? Boolean(opts.deploymentId && opts.apiKey);
    if (wantTelemetry && opts.deploymentId && opts.apiKey) {
      runner.reporter = new TelemetryReporter(
        resolveBaseUrl(opts.baseUrl),
        opts.deploymentId,
        opts.apiKey,
        {
          sdkVersion: VERSION,
          enabled: true,
          activeProvider: active,
          ortVersion: ortVersion(),
          availableProviders: opts.providers ?? [active],
        },
      );
    }
    return runner;
  }

  get activeProvider(): string {
    return this.activeProviderName;
  }

  /** The model's input specs (name, dims, type) — handy for building probes. */
  inputSpecs(): InputMeta[] {
    return this.metas.map((m) => ({
      name: m.name,
      dims: m.dims ? [...m.dims] : undefined,
      type: m.type,
    }));
  }

  private readInputMetas(): InputMeta[] {
    const meta = (this.session as { inputMetadata?: ReadonlyArray<OrtValueMetadata> })
      .inputMetadata;
    if (Array.isArray(meta) && meta.length) {
      return meta.map((m) => ({
        name: m.name,
        dims: normDims(m.dimensions ?? m.shape),
        type: normType(m.type),
      }));
    }
    return this.session.inputNames.map((name) => ({ name }));
  }

  async run(inputs: RunInput = null, opts: { region?: string } = {}): Promise<RunOutput> {
    const region = opts.region ?? "local";
    const t0 = performance.now();
    let feeds: Record<string, OrtTensor> = {};
    try {
      feeds = this.prepare(inputs);
      const preMs = performance.now() - t0;

      const t1 = performance.now();
      const result = await this.session.run(feeds);
      const inferMs = performance.now() - t1;

      const t2 = performance.now();
      const outputs: Array<{ name: string; shape: number[] }> = [];
      const raw: OrtTensor[] = [];
      for (const name of this.session.outputNames) {
        const t = result[name];
        if (!t) continue;
        raw.push(t);
        outputs.push({ name, shape: [...t.dims] });
      }
      const first = raw[0] ?? null;
      this.reporter?.observe(
        feedsToSampleable(feeds),
        first ? { data: first.data as ArrayLike<number | bigint>, dims: first.dims } : null,
      );
      const postMs = performance.now() - t2;

      this.reporter?.recordEvent({
        latencyMs: inferMs,
        preMs,
        postMs,
        success: true,
        batchSize: batchOf(feeds),
        region,
        inputSig: signatureOf(feeds),
      });
      return {
        latencyMs: round3(inferMs),
        preMs: round3(preMs),
        postMs: round3(postMs),
        outputs,
        raw,
      };
    } catch (err) {
      const errorCode = err instanceof BadInputError ? "bad_input" : "inference_error";
      this.reporter?.recordEvent({
        latencyMs: 0,
        success: false,
        errorCode,
        batchSize: batchOf(feeds),
        region,
        inputSig: signatureOf(feeds),
      });
      throw err;
    }
  }

  private prepare(inputs: RunInput): Record<string, OrtTensor> {
    if (inputs == null) {
      return randomProbe(this.ort, this.metas);
    }
    const feeds: Record<string, OrtTensor> = {};
    for (const name of this.session.inputNames) {
      const userInput = inputs[name];
      if (!userInput) throw new BadInputError(`missing input '${name}'`);
      feeds[name] = buildTensor(this.ort, userInput);
    }
    return feeds;
  }

  async close(): Promise<void> {
    await this.reporter?.close();
    try {
      await (this.session as { release?: () => Promise<void> }).release?.();
    } catch {
      /* best effort */
    }
  }
}

const round3 = (x: number): number => Math.round(x * 1000) / 1000;

function feedsToSampleable(feeds: Record<string, OrtTensor>): Record<string, Sampleable> {
  const out: Record<string, Sampleable> = {};
  for (const [k, t] of Object.entries(feeds)) {
    out[k] = { data: t.data as ArrayLike<number | bigint>, dims: t.dims };
  }
  return out;
}

function normDims(shape: ReadonlyArray<number | string> | undefined): number[] | undefined {
  if (!shape) return undefined;
  // Symbolic/dynamic dims (strings, <=0) become -1; randomProbe maps them to 1.
  return shape.map((d) => (typeof d === "number" && Number.isInteger(d) && d > 0 ? d : -1));
}

function normType(type: string | undefined): string | undefined {
  if (!type) return undefined;
  // ORT metadata types look like "tensor(float)" / "tensor(int64)".
  const m = /tensor\(([^)]+)\)/.exec(type);
  const t = (m ? m[1] : type)!.toLowerCase();
  switch (t) {
    case "float":
      return "float32";
    case "double":
      return "float64";
    default:
      return t;
  }
}
